### ISSUE TITLE
Syntax for Table Heatmap

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#1342]
Currently heatmap can be created using a `bar` mark.  

```
{
  "description": "A bar chart showing the US population distribution of age groups in 2000.",
  "data": { "url": "data/population.json"},
  "mark": "bar",
  "encoding": {
    "y": {
      "field": "age", "type": "ordinal"
    },
    "x": {
      "field": "year", "type": "ordinal"
    },
    "color": {
      "aggregate": "mean", "field": "people", "type": "quantitative" 
    }
  }
}
```

but that might not be the most natural way to do it.  We might want to consider other alternative. 

@willium's thought (copied from #1339)

> I'm also curious to see if the semantics of a heat map could be distanced from the "bar" mark too -- as discussed. It would not be my first guess, if I was trying to implement a heat map in Vega-lite. While troublesome, some sort of table seems more reasonable to me. What about the introduction of a "cell" mark (i.e. table-cell)?

**Comment by kanitw** [comment]
For comparison, Tableau has a weird semantic for its `square` marks for this case. 

Meanwhile, ggplot has a more sensible [`tile` marks](http://docs.ggplot2.org/current/geom_tile.html).

**Comment by willium** [comment]
I like tile as well (though cell might be more flexible if we end up wanting to support tables? It's also more ambiguous.

**Comment by willium** [comment]
https://vega.github.io/vega-editor/?mode=vega-lite&spec=text_table_heatmap is related

**Comment by kanitw** [comment]
Let's go with tile.   We still have to consider how to deal with tile for quantitative field though.

**Comment by kanitw** [comment]
Another alternative to consider is `rect`.  

Basically, if there is only `x` (but not `x2`) we can treat `x` as `xc` and expand the `rect` along the x-axis.  And if both `x` and `x2` exist, then this behaves like rect in Vega.  This would be consistent with what @ayush29feb is doing.

**Comment by pierre-haessig** [comment]
Hi,

I just started exploring vega-lite (through the use of altair), and I quickly fell on the issue of how to draw heatmap (which in my particular use case happens to be just a 2d histogram/density plot). Unless I missed something, the syntax using `bar` mark is not advertised in the doc, rather only present in the issue tracker.

On the other hand, the example gallery provides the "text table heatmap" (which in turns is reused in [Altair Plot Recipes](https://altair-viz.github.io/recipes.html#heatmaps)), but I feel this plot is strangely repurposing the row/column encoding channels, rather than using the regular x/y. I feel that the existence of this example gives a sense of confusion in the API.

My proposal is too add an new heatmap example in the doc gallery. I've started with the [Binned Scatterplot](https://vega.github.io/vega-editor/?mode=vega-lite&spec=scatter_binned) example, since I think this plot is really an alternative viz for a density plot. The result is almost the same, except for the mark and the third channel (`size` replaced by `color`).

```json
{
  "data": {"url": "data/movies.json"},
  "mark": "bar",
  "encoding": {
    "x": {
      "bin": {"maxbins": 10},
      "field": "IMDB_Rating",
      "type": "quantitative"
    },
    "y": {
      "bin": {"maxbins": 10},
      "field": "Rotten_Tomatoes_Rating",
      "type": "quantitative"
    },
    "color": {
      "aggregate": "count",
      "field": "*", "type":
      "quantitative"
    }
  }
}
```

![vega](https://cloud.githubusercontent.com/assets/1010404/19686025/bef31c1c-9abf-11e6-9a95-8608408f0b80.png)

**Comment by domoritz** [comment]
Thanks for the example. Until we decide on a better mark, we should use your example on the website.

**Comment by pierre-haessig** [comment]
Thanks for pushing this in the code.

**Comment by willium** [comment]
I dig this example. Looking forward to a mark for this as well! Surprisingly, I find myself needing heat maps frequently. Some of Vega 3.0's scene-graph stuff might make this even better :)

**Comment by aishfenton** [comment]
+1 for `tile`.

**Comment by willium** [comment]
I'd prefer `cell` because of the possibility of using text? I'd like to be able to compose heat tables with Vega-Lite. It's something I do several times a week

**Comment by g3o2** [comment]
+1 for `rect`, for that's exactly what it looks like and it would be fully consistent with vega lingo

**Comment by willium** [comment]
I feel like it's a little too close to vega if it's rect

**Comment by kanitw** [comment]
It will be a rect as rect works with both quantitative and ordinal scales.
Tile sounds more specific to ordinal scales on both axes.

I dont think its a big problem to be close with vega.

**Comment by aishfenton** [comment]
My thinking against rect was that it doesn't describe what it does. What
makes this mark different from the existing "square" is that it fills the
cell completely.

Rect to me sounds like it should be like square but let you specify a width
and height. It doesn't seem obvious to me that this'd be the one to use for
heat maps.

**Comment by willium** [comment]
I disagree. Heat maps are often rectangular (consider two ordinal scales, one much longer than the other). I dislike text because of potential to use text, however will concede that this would still theoretically be possible with reactive geometry... except with varied string lengths (which I think is a possibility, though less so

**Comment by kanitw** [comment]
I dont see why it doesnt describe what it does.
You are obviously drawing rectangles.

Square is what tableau uses for (table) heatmap, but heatmap can have x/y
scales with different bandwidth, which is by definition not squares.

If you re not prior ggplot2 users, rect is more obvious (match well with
vega, svg, and many other 2d graphics API).

**Comment by kanitw** [comment]
That said, to avoid confusion, we might remove squares and circles as you
can use point to do them. @domoritz

**Comment by aishfenton** [comment]
Ok. I see what you mean now. My main use of them is heatmaps of matrix
data, so they're always square (and usually only a pixel big). But I see
what you mean for the general case.

**Comment by kanitw** [comment]
Fixed in #1627 (pending review)
